src/charm.py

```diff
@@ -4,101 +4,424 @@
+from ops.charm import CharmBase, RelationEvent
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, MaintenanceStatus, Relation, WaitingStatus
+from ops.pebble import PathError
 
 logger = logging.getLogger(__name__)
 
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
+_TRAEFIK_CONTAINER_NAME = "traefik"
+_TRAEFIK_LAYER_NAME = "traefik"
+_TRAEFIK_SERVICE_NAME = "traefik"
+# We watch the parent folder of where we store the configuration files,
+# as that is usually safer for Traefik
+_TRAEFIK_INGRESS_CONFIGURATIONS_DIRECTORY = "/opt/traefik/juju"
 
-    _stored = StoredState()
+
+class TraefikIngressCharm(CharmBase):
+    """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+        self._port = 80
+        self._diagnostics_port = 8082
+
+        self.service_patch = KubernetesServicePatch(
+            charm=self,
+            service_type="LoadBalancer",
+            ports=[(f"{self.app.name}", self._port)],
+        )
+
+        self.metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{self._diagnostics_port}"]}],
+                },
+            ],
+        )
+
+        self.framework.observe(self.on.traefik_pebble_ready, self._on_traefik_pebble_ready)
+
+        # TODO Need to ensure K8s service has LoadBalancer IP before processing proxy requests?
+
+        self.framework.observe(self.on["ingress"].relation_created, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_joined, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_changed, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_departed, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_broken, self._handle_ingress_broken)
+
+    def _on_traefik_pebble_ready(self, event):
+        # Ensure the required basic configurations and folders exist
+        container = self.unit.get_container(_TRAEFIK_CONTAINER_NAME)
+
+        # TODO Handle case the relation events are deferred and this hook has not
+        # run yet when the pebble event queueing triggers the previously deferred
+        # events.
+        if not container.can_connect():
+            self.unit.status = WaitingStatus(
+                f"container '{_TRAEFIK_CONTAINER_NAME}' not yet ready"
+            )
+            return
```

Comment:
`event.defer()` before this? To be honest, you can probably drop this check in general -- this event is fired precisely *because* Juju can connect to it... leaving this in isn't *wrong* mind you